### ISSUE TITLE
increasing num_image_per_prompt from 16 to 24 leads to bad images

### ISSUE THREAD

**Issue by garychan22** [yifan123/flow_grpo#92]
Hi, thanks for your excellent work.

I am trying to fine-tune the diffusion model with two reward models. At the beginning, I used config.sample.num_image_per_prompt=16 to see if the reward model works, and found everything is fine but the generated images are not changed significantly. Then, I used config.sample.num_image_per_prompt=24 for improved performances. At the similar iterations, the run with config.sample.num_image_per_prompt=24 produced bad images including grid pattern and blurness. And I observed that one of the rewards dropped while the other one increased suddenly and fast.

Since I am new to grpo/rl, I really don't know why.

Appreciate any help or guidance, thanks.

**Comment by garychan22** [comment]
> Thank you very much for your interest in our work. The first question is whether setting `num_image_per_prompt = 24` would yield a higher overall reward compared to 16. Secondly, the drop in image quality is due to reward hacking; you need to increase `config.train.beta` to prevent the model parameters from changing too much.

Thanks for your reply! 

Setting ``num_image_per_prompt = 24`` yielded a lower overall reward compared to 16 in my cases. Dose this mean that the effect of the reward is amplified with larger group, then the reward hacking appears more faster (in early iterations) with small ``config.train.beta = 0.001`` (the fact may be that  the reward hacking will still appear when ``num_image_per_prompt = 16`` but in later iterations with small ``config.train.beta = 0.001``)?

Is increasing ``config.train.beta`` expected to observe significant visual improvement with more training steps?

Since I only have two nodes with 8 gpus each, I want to do fast evaluation on the effect of different reward models. For training speed up, should I be careful about decreasing ``config.train.timestep_fraction`` from 0.99 to 0.6 or using the new flow_grpo_fast with fine-grained image reward models (such as image quality assessment model, since the visual differences measured by these models usually appear in later denosing steps)?

Thanks!

**Comment by yifan123** [comment]
I highly recommend using **Flow-GRPO-Fast**, which provides more than a 2× speedup and achieves higher rewards. I suggest setting **beta = 0.01**.

**Comment by garychan22** [comment]
@yifan123 As suggested, I am trying to adapte the Flow-GRPO-Fast to my model, and found in https://github.com/yifan123/flow_grpo/blob/main/flow_grpo/diffusers_patch/sd3_pipeline_with_logprob_fast.py#L132-L134
```
    random.seed(process_index)
    if random_timestep is None:
        random_timestep = random.randint(0, sample_num_steps//2)
```
Dose this enforce that each group for a unique prompt shares the same timestep range for sde sampling? If the batch size is 2, then the two groups share the same timestep range for sde sampling?

Since I am using fsdp instead of multigpu with accelerate, should I use the batch_index instead of process_index to maintain this consistency?

Thanks!

**Comment by garychan22** [comment]
@yifan123 Thanks for the detailed explanations! I think my problem has been addressed.
